### ISSUE TITLE
Performance Regression 2 Recursive Types

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#3014]
**Describe the bug**
This is very similar to previous performance regression issue I made. This one comes from a different commit though (although maybe like last one real issue is separate) and latest version still shows slow performance. From 1.1.212 to 1.1.213 there was a performance regression for this code,

```python
from collections import abc
from functools import partial
from typing import Callable, Dict, List, Mapping, Sequence, TypeVar, Union

T1 = TypeVar("T1")
T2 = TypeVar("T2")

ContainerGeneric = Union[
    Mapping[str, "ContainerGeneric[T1]"], Sequence["ContainerGeneric[T1]"], T1
]
ContainerGeneric2 = Union[str, int, bool, None, Dict[str, "ContainerGeneric2"], List["ContainerGeneric2"]]


def container_fmap(
    f: Callable[[T1], T2], elements: ContainerGeneric[T1]
) -> ContainerGeneric[T2]:
    if isinstance(elements, abc.Mapping):
        return {key: container_fmap(f, val) for key, val in elements.items()}
    if isinstance(elements, tuple):
        return tuple(map(partial(container_fmap, f), elements))
    if isinstance(elements, abc.Sequence) and not isinstance(elements, (str, bytes)):
        return [container_fmap(f, elem) for elem in elements]
    return f(elements)


def replace_templates(template_mapping: Mapping[str, str], value: T1) -> T1:
    if isinstance(value, str):
        for template, replacement in template_mapping.items():
            if template in value:
                return value.replace(template, replacement)
    return value


def update_config(config: ContainerGeneric2) -> ContainerGeneric2:
    return container_fmap(lambda value: replace_templates({}, value), config)  # type: ignore

```

On 1.1.212 it took 1 second. On 1.1.213 it took 8-10 seconds.

**VS Code extension or command-line**
I also tested latest pyright (1.1.219) and several versions in between 1.1.213. The issue is present in 1.1.217-1.1.219 so looks like fix for that issue doesn't cover this case. 

1.1.212 - 1 second
1.1.213 - 8 seconds
1.1.217/1.1.218 - 6 seconds
1.1.219 - 10 seconds

edit: I just noticed 1.1.220 appeared. Checking that one also gives about 10 seconds.

**Comments**

The key line that looks to influence runtime is ContainerGeneric2. If change ContainerGeneric2 to instead use Mapping/Sequence it becomes fast again. 

```python
ContainerGeneric2 = Union[str, int, bool, None, Mapping[str, "ContainerGeneric2"], Sequence["ContainerGeneric2"]]
```
is fast

```python
ContainerGeneric2 = Union[str, Dict[str, "ContainerGeneric2"], List["ContainerGeneric2"]]
```
is also fast, but
```python
ContainerGeneric2 = Union[str, int, bool, None, Dict[str, "ContainerGeneric2"], List["ContainerGeneric2"]]
```
is slow.

On positive side most of my recursive type usage has better performance now and this particular function is rare. So I'm fine adjusting my code and just using a wrong type in rare spots where this slowness appears.

**Comment by huntzhan** [comment]
Observe the similar performance regression in `1.1.220`. After rolling back to `1.1.215` this issue is mitigated.

**Comment by erictraut** [comment]
The problem here involves type compatibility checks between two distinct recursive type aliases. My previous optimization addressed the case where both types were the same recursive type alias. I've generalized the optimization to handle this case as well. This will be included in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.221, which I just published. It will also be included in the next release of pylance.
